go/consensus/tendermint/abci/prune_test.go

```diff
@@ -1,36 +1,77 @@
 package abci
 
 import (
+	"context"
 	"fmt"
+	"io/ioutil"
+	"os"
 	"testing"
 
 	"github.com/stretchr/testify/require"
-	"github.com/tendermint/iavl"
-	dbm "github.com/tendermint/tm-db"
+
+	"github.com/oasislabs/oasis-core/go/common"
+	"github.com/oasislabs/oasis-core/go/common/crypto/hash"
+	mkvs "github.com/oasislabs/oasis-core/go/storage/mkvs/urkel"
+	mkvsDB "github.com/oasislabs/oasis-core/go/storage/mkvs/urkel/db/api"
+	mkvsBadgerDB "github.com/oasislabs/oasis-core/go/storage/mkvs/urkel/db/badger"
 )
 
 func TestPruneKeepN(t *testing.T) {
-	db := dbm.NewMemDB()
-	tree := iavl.NewMutableTree(db, 128)
-
-	for i := int64(1); i <= 11; i++ {
-		tree.Set([]byte(fmt.Sprintf("key:%d", i)), []byte(fmt.Sprintf("value:%d", i)))
-		_, ver, err := tree.SaveVersion()
-		require.NoError(t, err, "SaveVersion: %d", i)
-		require.Equal(t, i, ver, "incorrect version on save")
+	require := require.New(t)
+
+	// Create a new random temporary directory under /tmp.
+	dir, err := ioutil.TempDir("", "abci-prune.test.badger")
+	require.NoError(err, "TempDir")
+	defer os.RemoveAll(dir)
+
+	// Create a Badger-backed Node DB.
+	ndb, err := mkvsBadgerDB.New(&mkvsDB.Config{
+		DB:           dir,
+		NoFsync:      true,
+		MaxCacheSize: 16 * 1024 * 1024,
+	})
+	require.NoError(err, "New")
+	tree := mkvs.New(nil, ndb)
+
+	ctx := context.Background()
+	for i := uint64(1); i <= 11; i++ {
+		err = tree.Insert(ctx, []byte(fmt.Sprintf("key:%d", i)), []byte(fmt.Sprintf("value:%d", i)))
+		require.NoError(err, "Insert")
+
+		var rootHash hash.Hash
+		_, rootHash, err = tree.Commit(ctx, common.Namespace{}, i)
+		require.NoError(err, "Commit")
+		err = ndb.Finalize(ctx, common.Namespace{}, i, []hash.Hash{rootHash})
```

Comment:
In case of Tendermint we could merge them (as there is always only one root per round which is the final root) to reduce some extra db round-trips. But most of those round-trips will be to in-memory memtables so it shouldn't account for much while a special case would add some code.

The reason why these two are separate in our MKVS is because we support the runtime state updates use case where multiple compute nodes can commit multiple roots. But only one is eventually finalized.

Reply:
oh thanks for the explanation